### ISSUE TITLE
Reading Nexus file with datatype=standard gets symbols wrong

### ISSUE THREAD

**Issue by BalduinLandolt** [biopython/biopython#2814]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
>>> import sys; print(sys.version)
3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:05:16) [MSC v.1915 32 bit (Intel)]
>>> import platform; print(platform.python_implementation()); print(platform.platform())
CPython
Windows-10-10.0.18362-SP0
>>> import Bio; print(Bio.__version__)
1.76
```

### Expected behaviour
Loading a nexus file... hopefully without error.

### Actual behaviour
Getting an error when loading a nexus file.
```
Traceback (most recent call last):
  File "C:/Users/Balduin Landolt/switchdrive/Documents_Balduin/Uni/@aktuel/Seminararbeit - Balladen/biopython_test/bio-test.py", line 21, in <module>
    run_test()
  File "C:/Users/Balduin Landolt/switchdrive/Documents_Balduin/Uni/@aktuel/Seminararbeit - Balladen/biopython_test/bio-test.py", line 7, in run_test
    alignment = AlignIO.read("trivial.nex", "nexus")
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\__init__.py", line 447, in read
    first = next(iterator)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\__init__.py", line 394, in parse
    for a in i:
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\NexusIO.py", line 42, in NexusIterator
    n = Nexus.Nexus(handle)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Nexus\Nexus.py", line 682, in __init__
    self.read(input)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Nexus\Nexus.py", line 732, in read
    self._parse_nexus_block(title, contents)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Nexus\Nexus.py", line 773, in _parse_nexus_block
    getattr(self, "_" + line.command)(line.options)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Nexus\Nexus.py", line 1109, in _matrix
    % (id, coding, iupac_seq)
Bio.Nexus.Nexus.NexusError: Taxon _text_05__: Illegal character 2 in sequence 2211123?12-24 (check dimensions/interleaving)
```

### Steps to reproduce
Following Nexus file:
```
#NEXUS file "trivial.nex"
BEGIN data;[eröffnet den Data-Block]
Dimensions ntax=5 nchar=13; [Definiert die Größe des Alignments]
Format
datatype=standard
missing=?
gap=- [Definiert den Datentyp und Symbole für fehlende Daten (?) und gaps (-)]
Symbols = "0 1 2 3 4 5 6 7 8 9";
Matrix [hier beginnt das Alignment...]
_text_05__ 2211123?12-24
_text_04__ 3212424112223
_text_03__ 2-22313122-22
_text_02__ 1122212111-21
_text_01__ 11111?1111111
; [...und hier endet es]
END; [beendet den Data-Block]
```
(never mind the comments)

Running the following python script:

```python

from Bio import AlignIO


def run_test():
    alignment = AlignIO.read("trivial.nex", "nexus")
    print(alignment)


if __name__ == '__main__':
    run_test()
```

Instead of the output, I get the above added error stack trace.

I did look quite a bit into it already:  
The error is raised [here](https://github.com/biopython/biopython/blob/fdd4c7b51ebfb28933a9b803745bd9257d7a2150/Bio/Nexus/Nexus.py#L1091-L1101).  
At that point, the nexus object has `self.symbols = ['"', '0']`, so it's obvious, that other values mus cause a crash.

This seems to be caused [here](https://github.com/biopython/biopython/blob/fdd4c7b51ebfb28933a9b803745bd9257d7a2150/Bio/Nexus/Nexus.py#L819-L836), by the `options` variable that has `{'datatype': 'standard', 'missing': '?', 'gap': '-', 'symbols': '0', '1': None, '2': None, '3': None, '4': None, '5': None, '6': None, '7': None, '8': None, '9': None}`.  
Clearly, the symbols get parsed wrong.
This seems to happen [here](https://github.com/biopython/biopython/blob/fdd4c7b51ebfb28933a9b803745bd9257d7a2150/Bio/Nexus/Nexus.py#L596-L617), where `line = 'Format datatype=standard missing=? gap=-  Symbols=0 1 2 3 4 5 6 7 8 9'`is parsed into `options = ['datatype', '=', 'standard', 'missing', '=', '?', 'gap', '=', '-', 'Symbols', '=', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9']`

I hope this info helps!

**Comment by peterjc** [comment]
Does removing the whitespace in the symbols line work? ie. ``Symbols = "0 1 2 3 4 5 6 7 8 9";`` to ``Symbols = "0123456789";``

(I would try it myself but a little tied up right now with other things)

**Comment by BalduinLandolt** [comment]
Indeed it does... to some extent, at least. I now get a different error message:
```
Traceback (most recent call last):
  File "C:/Users/Balduin Landolt/switchdrive/Documents_Balduin/Uni/@aktuel/Seminararbeit - Balladen/biopython_test/bio-test.py", line 21, in <module>
    run_test()
  File "C:/Users/Balduin Landolt/switchdrive/Documents_Balduin/Uni/@aktuel/Seminararbeit - Balladen/biopython_test/bio-test.py", line 7, in run_test
    alignment = AlignIO.read("trivial.nex", "nexus")
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\__init__.py", line 447, in read
    first = next(iterator)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\__init__.py", line 394, in parse
    for a in i:
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\AlignIO\NexusIO.py", line 63, in NexusIterator
    yield MultipleSeqAlignment(records, n.alphabet)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Align\__init__.py", line 169, in __init__
    self.extend(records)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Align\__init__.py", line 523, in extend
    self._append(rec, expected_length)
  File "C:\Program Files (x86)\python\lib\site-packages\Bio\Align\__init__.py", line 597, in _append
    if not Alphabet._check_type_compatible([self._alphabet, record.seq.alphabet]):
AttributeError: 'StandardData' object has no attribute 'alphabet'
```

I'll digg into what that error means as soon as I find time.

But in any case, valid Nexus could be either with wite space or without, as far as I know. So that might need fixing anyway.  
See e.g. [Maddison et al. 1997](https://watermark.silverchair.com/46-4-590.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAmAwggJcBgkqhkiG9w0BBwagggJNMIICSQIBADCCAkIGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMIzXL870DN4gmdZy8AgEQgIICE9bTOTySWI6Uz2z6DMlU0tDvPvexT--Zb78Wh2-w3YjrMn5kPyRpMUgL5GAbKJBgmKERwnMNHPyUmGmEY-6V0GywSIeo_WkkMMaX9h_p_SuexHljLymL2bMOftIPo6vvyilBh7FIurlhksjhwWA-pFzg382nUdNqQFNTawgFUDYTia5jzQqsOZVCkn8j2uFb5iKZNenxTFtdmltN8XQGaxMfRN9MTe1ocwUNEzGE6pV68W8NhvxeAB8QJ3z_ZwozqbUtkisQ__Q7gUtB6J36h4__MBvnA__iZGDEw3z6ojVJQi3X2atl7jXAHA_RntNbHQ0i22eaR2RWA_GNoZYklMCom6U9boO5oK8C8g_5jO12-nnwQRtpNn4wVWdkVi18dksXvxxOTv6tJ1WW_vKwnHBXydTpvuxJDQ28itND2szcp8OL1qicI615zRUeUqe48jlLesQbUh32ESa7MQD-z7lWJ1DOWSTCXcGYdrQqxXWVt8RI9qdMS_S1u-S6UMNkAnfWQswJnMuSuSLbKJcu-o95agvva63UZMu6T1U4G1meJ49v1B1P5LXFEqB4VSX-iRcl21XpxmBUD2WVgdHDrTaesOYz9BDwM1MJi0xPVYBw6X4zzU3Dor70Us7C7PJ_gADe3Rgy3OMs4v6XPWiORgcDUwLRj-0k7tJeIwBn_ivgLmRGE-JozAvCchN5KsglNsy2vQ), p. 599:  
"For example, SYMBOLS="01234567" designates numbers 0 through 7 as acceptable symbols in a matrix. The SYMBOLS subcommand is not allowed for DATATYPE=CONTINUOUS. The default symbols list differs from one DATATYPE to another, as described understate symbol in the Appendix.  Whitespace is  not needed between elements: SYMBOLS="012" is equivalent to SYMBOLS="0 1 2"."  
So basically both should work.

**Comment by peterjc** [comment]
The comment here https://github.com/biopython/biopython/blob/fdd4c7b51ebfb28933a9b803745bd9257d7a2150/Bio/Nexus/Nexus.py#L819-L836 was why I suggested ``Symbols = "0123456789";``

That quote does suggest either should work. As to the next error, ``Bio.AlignIO`` has only been extensively tested with RNA/DNA/Protein sequences, so trying digits 0-9 is going to be uncharted territory. Happily there are all single characters, so it could be made to work.

If you stick with ``Bio.Nexus`` only you might be better off. What is your usecase?

**Comment by peterjc** [comment]
It looks like line 820 as quoted, ``self.alphabet = None`` could to be something like:

 ```python
    self.alphabet = SingleLetterAlphabet()
    self.alphabet.letters = self.symbols
```

(Or we do something similar in ``Bio/AlignIO/NexusIO.py`` when the alphabet is ``None``)

With that change the example gets further, but fails with ``Bio.Nexus.Nexus.NexusError: Unmatched 'end'`` which I think is just because this is a partial example.

Can you share a complete NEXUS file for testing, ideally one we could include in future Bioython releases as part of our test suite?

**Comment by BalduinLandolt** [comment]
Well... true... I should have spotted that comment myself. No idea how I missed that.  
(Though it probably shouldn't be the case.)

As to my use case: I work with humanities data, mostly medieval manuscripts, but also folklore motivs.  
For now I'm really just fiddling around and testing stuff. But what I have in mind (and people have done that before) is taking related artefacts (say different manuscripts of the same text) and break them down into small segments, and compare the segment among all witnesses; the first one gets assigned 0, if the second one is the same, it gets assigned 0 as well, if the third one is different, it gets assigned 1, and so forth. Like this I can get an alignment from which I want to infer the "phylogeny" of my manuscripts.

For now, I'm really still testing stuff, so the nexus file I pasted above is all I have.  
Correct me if I'm wrong, but as far as I know (again, I'm not a biologist), this should be complete and valid nexus. I basically copied it from the [german Wikipedia here](https://de.wikipedia.org/wiki/Nexus_(Bioinformatik)) and replaced the DNA data with numbers.  
A while back, I also got SplitsTrees to parse that file; and importing it to R worked fine too. (But who wants to work with R when there is Python...)

I'm of course more than willing to share a file that is a bit more sensible, as soon as I have actual data to work with. But for now, this is all I have.

Also, I only wanted to bring this potential bug to your attention. It's not vital for me to have it fixed:  
I created a Phylip-file from the same data and that parsed without issue (it even gave me a nice little dendrogram :) ), so I can work with that for now.

In any case, thank you for your help! And thanks to all of you guys for this package!

**Comment by peterjc** [comment]
Cool use case, a little outside our norm, but relatable. If you had more than 10 categories, would some other single letter symbols have worked fine (e.g. 0 to 9 and then letters like in hexadecimal notation)?

I think this can be made to work, but if you can share any other examples which work in other tools, that would be helpful - Wikipedia licensing isn't ideal for reuse in a software package.

I personally have minimal experience with the Nexus format - but did write the AlignIO code which sits on top of it. I should have tried a Google translate ("und hier endet es" = "and here it ends"). Given the error message, I wrongly guessed that was something like "and more like this" indicating a partial example.

**Comment by BalduinLandolt** [comment]
Glad to hear you don't find my use case entirely absurd. :)

Generally, yes, after 9 one could continue with hexadecimal notation - and even further: 0-z would already make 36 states (and then one could test if all non-whitespace, non-punctuation, non-combining unicode characters work... which would leave me with more than enough.  
But in practice I'd probably always try to avoid using more than 10 states. If a position has more than a couple of significant variants, that usually indicates that the segmentation is not fine-grained enough and that some of the variants are entirely independent of one another. In these cases, I'd split the segment into two; and these segments would get numbered independently.  
Here's an example to illustrate this: If we have three text witnesses:
```
"The man walks over the field"
"The man walks through the forest"
"The woman walks over the field"
```
It would be wrong to interpret this as three independent variants:
```
"The man walks over the field"
 1
"The man walks through the forest"
 2
"The woman walks over the field"
 3
```
Much rather, this should be seen as two segments:
```
"The man      |    walks over the field"
 1                 1
"The man      |    walks through the forest"
 1                 2
"The woman    |    walks over the field"
 2                 1
```
So the alignment would be
```
11
12
21
```
instead of
```
1
2
3
```
Obviously, neither of those produce a meaningful phylogenetic tree... But you get the point. :)

I'll gladly share a more meaningful and license-free example , as soon as I have some data. Unfortunately - but as it tends to be all the time - more pressing matters keep popping up.  
Should we just leave the issue open, so I can post the example here, as soon as I have it? Or what works best for you?

**Comment by peterjc** [comment]
That sounds good - leave this open pending some test data and some time to fix this properly. Thank you!
